File: python/thamudic/old_north_arabian.py
"""Canonical Old North Arabian (Ancient North Arabian) Unicode registry.

The Unicode repertoire is U+10A80..U+10A9F. Unicode encodes the repertoire
using Dadanitic forms; Safaitic, Hismaic, Taymanitic, Minaic and Thamudic B
have historically attested variant glyph forms that should be represented by
font/variant metadata rather than invented Unicode code points.
"""

FIRST, LAST = 0x10A80, 0x10A9F
VARIANT_FORMS = ("Dadanitic", "Safaitic", "Hismaic", "Taymanitic", "Minaic", "Thamudic B")
_NAMES = [
    "HEH", "LAM", "HAH", "MEEM", "QAF", "WAW", "ES-2", "REH", "BEH", "TEH", "ES-1", "KAF", "NOON", "KHAH", "SAD", "ES-3",
    "FEH", "ALEF", "AIN", "DAD", "GEEM", "DAL", "GHAIN", "TAH", "ZAIN", "THAL", "YEH", "THEH", "ZAH",
]
_TRANSLITERATION = ["h", "l", "ḥ", "m", "q", "w", "s2", "r", "b", "t", "s1", "k", "n", "ḫ", "ṣ", "s3", "f", "ʼ", "ʽ", "ḍ", "g", "d", "ġ", "ṭ", "z", "ḏ", "y", "ṯ", "ẓ"]

CHARACTERS = tuple({
    "codepoint": cp,
    "character": chr(cp),
    "name": f"OLD NORTH ARABIAN LETTER {name}",
    "transliteration": transliteration,
    "utf8": chr(cp).encode("utf-8"),
    "utf8_hex": chr(cp).encode("utf-8").hex(" ").upper(),
} for cp, name, transliteration in zip(range(FIRST, 0x10A9D), _NAMES, _TRANSLITERATION)) + tuple({
    "codepoint": cp,
    "character": chr(cp),
    "name": f"OLD NORTH ARABIAN NUMBER {label}",
    "transliteration": value,
    "utf8": chr(cp).encode("utf-8"),
    "utf8_hex": chr(cp).encode("utf-8").hex(" ").upper(),
} for cp, label, value in ((0x10A9D, "ONE", "1"), (0x10A9E, "TEN", "10"), (0x10A9F, "TWENTY", "20")))

BY_CODEPOINT = {item["codepoint"]: item for item in CHARACTERS}
BY_CHARACTER = {item["character"]: item for item in CHARACTERS}

def is_old_north_arabian(value: int | str) -> bool:
    cp = ord(value) if isinstance(value, str) else value
    return FIRST <= cp <= LAST

def utf8_bytes(value: int | str) -> bytes:
    cp = ord(value) if isinstance(value, str) else value
    return chr(cp).encode("utf-8")

def transliterate(text: str) -> str:
    return "".join(BY_CHARACTER[ch]["transliteration"] if ch in BY_CHARACTER else ch for ch in text)

File: dotnet/src/Thamudic.Core/OldNorthArabian.cs
using System.Globalization;
using System.Text;
namespace Chimera.Thamudic;

public sealed record OldNorthArabianCharacter(int CodePoint, string Character, string Name, string Transliteration, string Utf8Hex);

public static class OldNorthArabian
{
    public const int First = 0x10A80;
    public const int Last = 0x10A9F;
    public static readonly string[] VariantForms = { "Dadanitic", "Safaitic", "Hismaic", "Taymanitic", "Minaic", "Thamudic B" };
    public static readonly IReadOnlyList<OldNorthArabianCharacter> Alphabet = Enumerable.Range(First, Last - First + 1)
        .Select(cp => new OldNorthArabianCharacter(cp, char.ConvertFromUtf32(cp), UnicodeName(cp), Transliteration(cp), Utf8Hex(cp)))
        .ToArray();

    public static bool IsOldNorthArabian(int codePoint) => codePoint >= First && codePoint <= Last;
    public static byte[] ToUtf8(int codePoint) => Encoding.UTF8.GetBytes(char.ConvertFromUtf32(codePoint));
    public static string Utf8Hex(int codePoint) => string.Join(" ", ToUtf8(codePoint).Select(b => b.ToString("X2", CultureInfo.InvariantCulture)));
    public static string Transliterate(int codePoint) => codePoint switch
    {
        0x10A80 => "h", 0x10A81 => "l", 0x10A82 => "ḥ", 0x10A83 => "m", 0x10A84 => "q", 0x10A85 => "w", 0x10A86 => "s2", 0x10A87 => "r",
        0x10A88 => "b", 0x10A89 => "t", 0x10A8A => "s1", 0x10A8B => "k", 0x10A8C => "n", 0x10A8D => "ḫ", 0x10A8E => "ṣ", 0x10A8F => "s3",
        0x10A90 => "f", 0x10A91 => "ʼ", 0x10A92 => "ʽ", 0x10A93 => "ḍ", 0x10A94 => "g", 0x10A95 => "d", 0x10A96 => "ġ", 0x10A97 => "ṭ",
        0x10A98 => "z", 0x10A99 => "ḏ", 0x10A9A => "y", 0x10A9B => "ṯ", 0x10A9C => "ẓ", 0x10A9D => "1", 0x10A9E => "10", 0x10A9F => "20", _ => ""
    };
    public static string Transliterate(string text) => string.Concat(text.Select(ch => IsOldNorthArabian(ch) ? Transliterate(ch) : ch.ToString()));
    private static string UnicodeName(int cp) => cp switch
    {
        <= 0x10A9C => $"OLD NORTH ARABIAN LETTER {new[]{"HEH","LAM","HAH","MEEM","QAF","WAW","ES-2","REH","BEH","TEH","ES-1","KAF","NOON","KHAH","SAD","ES-3","FEH","ALEF","AIN","DAD","GEEM","DAL","GHAIN","TAH","ZAIN","THAL","YEH","THEH","ZAH"}[cp-First]}",
        0x10A9D => "OLD NORTH ARABIAN NUMBER ONE", 0x10A9E => "OLD NORTH ARABIAN NUMBER TEN", 0x10A9F => "OLD NORTH ARABIAN NUMBER TWENTY", _ => ""
    };
}
